### PR TITLE
[Model] Add Granite Speech Support

### PR DESCRIPTION
**Pull request by alex-jw-brooks**
This PR adds support for Granite Speech models, and is a port of the corresponding [PR in Transformers](https://github.com/huggingface/transformers/pull/36801). The model uses a conformer-based encoder, with a blip2 qformer-based projector to encode the audio, and masks it into a granite LLM. This model also uses an audio-specific lora adapter, which should only be enabled when the model is processing audio inputs. Currently, this means that the user needs to make a `LoraRequest` every time they send audio. 

It is probably a good idea to wait for the transformers PR to be merged so that everything is aligned, but opening this PR in case anyone has feedback 🙂 unfortunately, a model compatible with this PR is not publicly available yet - I am happy to submit a follow-up PR adding an example / docs + tests once one it is out.

Some quirks that are good to be aware of / have kind of gross edge cases that I am actively looking into:

- The (rank 64) lora is bundled in the same dir as the model. At least in offline mode, it seems that the lora is loaded, but the lora layers are adding zero tensors, which result in unchanged outputs - still looking into this.

- The model is very sensitive - I haven't optimized the conformer implementation yet, but if possible, it would be great if we could for now avoid optimizing the conformer layers until we also have some tests for alignment with HF once the model is released, as the optimizations in the granite LLM already seem to shift things a bit, and I still need to run a quality benchmark (still looking into whatever is going on with the lora first!)

- Batching is a bit quirky because we don't use a feature attention mask and do zero padding _prior_ to calculating the Mel features in the HF processor (i.e., the padding indices end up at small negative numbers that are dependent on the batch, though they are masked out in transformers with a masked scatter which is the most important thing). Since the static batch is submitted one instance at a time to the processor in vLLM, this results in the features being unpadded; this PR handles this after the fact by zero padding the 3D Mel features and torch splitting the result, though maybe there is a better small negative value to use here.

CC @DarkLight1337 @njhill @tlrmchlsmth

### CODE REVIEW

**Comment by DarkLight1337** [comment]
Thanks for opening this!

> This model also uses an audio-specific lora adapter, which should only be enabled when the model is processing audio inputs. Currently, this means that the user needs to make a LoraRequest every time they send audio.

This is fine, it's somewhat like how Phi-4-multimodal is handled. Can you add this model to the examples so users know how to use it?

Also please update the supported models page and test registry.

**Comment by DarkLight1337** [comment]
Can you also add this model to `tests/models/multimodal/processing/test_common.py` to test the processing logic?

**Comment by DarkLight1337** [comment]
Also please add this model to the Supported Models doc page

File: vllm/model_executor/models/granite_speech.py [hunk 1]
```diff
@@ -0,0 +1,731 @@
...
+        inner_dim = config.dim_head * config.num_heads
+        self.max_pos_emb = config.max_pos_emb
+        self.context_size = config.context_size
+        self.num_heads = config.num_heads
+        self.dim_head = config.dim_head
+        self.scale = self.dim_head**-0.5
+        self.pre_norm = nn.LayerNorm(config.hidden_dim)
+        self.to_q = nn.Linear(config.hidden_dim, inner_dim, bias=False)
+        self.to_kv = nn.Linear(config.hidden_dim, inner_dim * 2, bias=False)
+        self.to_out = nn.Linear(inner_dim, config.hidden_dim)
```
Comment (DarkLight1337) [inline vllm/model_executor/models/granite_speech.py:286]: We should try to use the parallelized version of these layers where possible

**Comment by alex-jw-brooks** [comment]
Sorry for the ping before finishing the first requested changes, I think it may have automatically re-requested code owner review when I force pushed! That all sounds good to me, will work on it asap now that the transformers PR is merged

File: vllm/model_executor/models/granite_speech.py [hunk 2]
```diff
@@ -0,0 +1,731 @@
...
+        super().__init__()
+        self.hidden_size = config.projector_config.hidden_size
+        self.downsample_rate = config.downsample_rate
+        self.window_size = config.window_size
+        self.num_queries = config.window_size // config.downsample_rate
+
+        self.query = nn.Parameter(
+            torch.zeros(1, self.num_queries,
+                        config.projector_config.hidden_size))
+        self.query.data.normal_(mean=0.0, std=1.0)
```
[the remark this answers is not in the document]
Comment (alex-jw-brooks) [reply vllm/model_executor/models/granite_speech.py:210]: It shouldn't be! Will take a pass at removing things like this and switching out some of the linear layers to parallel versions in the next day or so

**Comment by alex-jw-brooks** [comment]
Hey @DarkLight1337, I think this should be ready for another look when you have a moment!

The bug fix for the lora name parsing https://github.com/vllm-project/vllm/pull/17196 is needed for this model to work properly, but things look aligned with the transformers PR when this PR is rebased on top of that one 🙂

File: tests/models/multimodal/processing/test_common.py [hunk 3]
```diff
@@ -285,6 +285,7 @@ def _test_processing_correctness_mistral(
     "Skywork/Skywork-R1V-38B",
     "fixie-ai/ultravox-v0_5-llama-3_2-1b",
     "openai/whisper-large-v3",
+    "ibm-granite/granite-speech-3.3-8b"
```
Comment (DarkLight1337) [inline tests/models/multimodal/processing/test_common.py:288]: Keep this in alphabetical order

File: vllm/model_executor/models/granite_speech.py [hunk 4]
```diff
@@ -0,0 +1,791 @@
...
+            hidden_states,
+            sampling_metadata,
+        )
+
+    def sample(
+        self,
+        logits: torch.Tensor,
+        sampling_metadata: SamplingMetadata,
+    ) -> Optional[SamplerOutput]:
+        return self.language_model.sample(logits, sampling_metadata)
```
Comment (DarkLight1337) [inline vllm/model_executor/models/granite_speech.py:776]: This method is not used anymore since #17084

File: vllm/model_executor/models/granite_speech.py [hunk 5]
```diff
@@ -0,0 +1,791 @@
...
+
+    def get_supported_mm_limits(self) -> Mapping[str, Optional[int]]:
+        return {"audio": 1}
+
+    def get_mm_max_tokens_per_item(
+        self,
+        seq_len: int,
+        mm_counts: Mapping[str, int],
+    ) -> Mapping[str, int]:
+        return {"audio": self.get_max_audio_tokens()}
```
Comment (DarkLight1337) [inline vllm/model_executor/models/granite_speech.py:81]: This method is not used anymore now. We can remove it

File: vllm/model_executor/models/blip2.py [hunk 6]
```diff
@@ -367,6 +367,7 @@ def __init__(
         *,
         quant_config: Optional[QuantizationConfig],
         cache_config: Optional[CacheConfig],
+        prefix: str = "",
```
[the remark this answers is not in the document]
Comment (alex-jw-brooks) [reply vllm/model_executor/models/blip2.py:380]: Yeah, it seems like blip2 only took it as an arg to the top level module and didn't pass it through - I went ahead and pushed the prefix through blip2 submodules as part of this PR!

File: vllm/model_executor/models/granite_speech.py [hunk 7]
```diff
@@ -0,0 +1,777 @@
...
+        prompt: str,
+        mm_data: Mapping[str, object],
+        mm_kwargs: Mapping[str, object],
+    ) -> BatchFeature:
+        mm_data = dict(mm_data)
+        audios = mm_data.pop("audios", [])
+
+        if audios:
+            # GraniteSpeechFeatureExtractor accepts "audio"
+            mm_data["audio"] = audios
```
Comment (DarkLight1337) [inline vllm/model_executor/models/granite_speech.py:148]: This looks unnecessary

**Comment by DarkLight1337** [comment]
Otherwise LGTM. Have you verified that the model works correctly on your end?

**Comment by alex-jw-brooks** [comment]
Yup, things look right on my side of things! I went ahead and pulled the audio assets fixtures from the ultravox tests to conftest and added a generation test under `tests/models/decoder_only/audio_language/test_granite_speech.py`, which won't currently run because transformers hasn't cut 4.52 yet, but it does pass on my machine using the tip of transformers.

Also realized the audio placeholder was missing for running online, so added that too 🙂

File: vllm/model_executor/models/granite_speech.py (hunk 7 above)
Comment (DarkLight1337) [reply vllm/model_executor/models/granite_speech.py:148]: Ah I see, thanks for the explanation

**Review by DarkLight1337 — APPROVED** [review]
LGTM then!
